src/compute/take/list.rs

```diff
@@ -0,0 +1,214 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use crate::{
+    array::{
+        growable::{Growable, GrowableList},
+        Array, ListArray, Offset, PrimitiveArray,
+    },
+    error::Result,
+};
+
+use super::maybe_usize;
+
+/// `take` implementation for ListArrays
+pub fn take<I: Offset, O: Offset>(
+    values: &ListArray<I>,
+    indices: &PrimitiveArray<O>,
+) -> Result<ListArray<I>> {
+    let usize_indices: Result<Vec<usize>> = indices
+        .values()
+        .iter()
+        .map(|i| maybe_usize::<O>(*i))
+        .collect();
+
+    let arrays: Vec<ListArray<I>> = usize_indices?
+        .into_iter()
+        .map(|index| values.slice(index, 1))
+        .collect();
```

Comment:
I think that this collect can also be removed, and `collect` the values directly to `let array_ref: Vec<&dyn Array> = ... .collect()`.

Reply:
I couldn't create a vector of references at the same time the slices are created. The complication is that the `array_ref` is made out of refences to the slices, which need to last longer than the references in `array_ref`. This is another reason why the Growable interface could be improved by passing the references when `extending` the array.

Reply:
I agree, I there are two problems imo. 1. `GrowableList` should expect a `Vec<&ListArray<O>>`, since:

* it takes ownership of the `vec`
* it downcasts arguments to `ListArray`

The first issue affects all interfaces, the latter issue is only for `GrowableList` and some others; `GrowablePrimitive` for example already expects a `PrimitiveArray`.

Great catch!

Reply:
I just pushed a change to the take function. With your suggestions I managed to remove some allocations and I managed to use iterators. However, I didn't manage to remove the loop to create the reference of arrays. As you pointed, this is needed for the actual implementation of Growable.

Reply:
Btw, If you change all the Growable implentations to take `Vec<&dyn Array>` instead of `&[&dyn Array]` the take implementation can be simplified further.

Reply:
Let's keep this PR to the `take` and handle this afterwards 👍